### ISSUE TITLE
Table multipolygons doesn't handle adjacent (closed) inner rings well: buildings on inner areas of farmland sharing sides (touching) w/other inner flagged as being on farmland.   

### ISSUE THREAD

**Issue by sekerob** [osm-fr/osmose-backend#2138]
Could not find the previous topic open or closed regarding buildings on farming lands, but today saw a building of a group on an 'inner' farmland residential area being flagged ([Verrière, France](https://www.openstreetmap.org/changeset/146506037#map=18/45.74904/0.23103) ). Farmland is defined as a multipolygon relation. This area shared a side with a farmland 'inner' wood area (is considered touching) i.e. 2x inner where before one would map a ring around the 2 touching areas and make that one inner. Was told on the community forum this is no longer needed. For the exercise, drew that ring and made it inner to see if the building(s) get flagged again. Both the residential and wood are simple polygons.

**Comment by Famlam** [comment]
I don't see any warnings related to buildings/farmlands in this region? (Note your changeset is only 5 hours old, but Osmose may take a day or so to update, so you may have seen an old marker)
https://osmose.openstreetmap.fr/nl/map/#zoom=17&lat=45.749629&lon=0.230487&item=xxxx&level=1%2C2%2C3&tags=&fixable=

Can you please give us:
1. The exact location
2. The item / class of the issue
3. Preferably the issue reported?
And as you mention a community discussion, please also link to the relevant (international) discussions. 

Note that the wiki of farmyard describes which buildings go on a farmyard rather than farmland. Farmland is really where the crops grow.
https://wiki.openstreetmap.org/wiki/Tag:landuse=farmyard

**Comment by sekerob** [comment]
".... so you may have seen an old marker)", not really since the item appeared this morning in the 'personal' trouble list. Stared for longer at this and checked all possibles but the areas were truly 'inner' to the proper farmland, The only difference seen from all else was this side touching with that adjacent wood which was also mapped as a direct inner.

Your wiki link "farmhouse, dwellings, farmsteads, sheds, stables, barns, equipment sheds, feed bunkers, etc" is pretty much all and everything, all have building=yes but I'm curious which exceptions of buildings there are not in 'etc' which one can map on farmland. What are the exceptions to that rule of having to be on an inner area?

Anyway, the link was provided, so if nothing shows tomorrow I'll deem it a confirmation of a bug.

The building flagged was this elongated up against the road. As said all buildiings on that residential patch are =yes

![image](https://github.com/osm-fr/osmose-backend/assets/23308504/edf78d08-11fe-440c-8e60-3f207991a39e)

**Comment by sekerob** [comment]
The same issue showed up this morning on 2 more 'inner' residential areas of farmland MPs (one residential hamlet named Le Breuil in the same zone linked in the OP, both with one of the many buildings on there being flagged. Both of these residential areas happened to share a side with another 'inner' area of the same farmland. So far not seen this on 'inner' areas that either are isolated/stand alone or are encapsulated in a single inner ring with other non-farmland areas. Just hit the OK on these buildings, not the false positive X and mapped encapsulated rings around them. If none of these many buildings get flagged again in the coming days, week, there's a case. Certainly worked to not have any other building lurking on these farmlands which are not put on an inner, mostly farmyards, dozens and dozens more.

**Comment by Famlam** [comment]
At this moment I still don't know what exactly the issue is, so please just share the error without modifying the data or closing the marker when they (re)appear :).

**Comment by sekerob** [comment]
Reproducible. Code 1310, level 3 on this building with pin, the residential area flanked by an 'inner' meadow' right top.
JOSM validator finds no wrong with the MP relation, nothing showing in OSM Inspector area view.

![image](https://github.com/osm-fr/osmose-backend/assets/23308504/fbde978b-a59c-4a51-b0ef-3692c1317e8f)
![image](https://github.com/osm-fr/osmose-backend/assets/23308504/4d83165a-1d8a-477b-8657-ce8208bd43cf)
![image](https://github.com/osm-fr/osmose-backend/assets/23308504/ca9cfa2a-fe98-48be-b67f-36d92f81f958)

**Comment by frodrigo** [comment]
Here https://osmose.openstreetmap.fr/fr/map/#zoom=16&lat=48.13858&lon=2.29591&item=1310&level=1%2C2%2C3

**Comment by Famlam** [comment]
Yep, it seems something is wrong with the building of multipolygons geometries if two inners are connected. This is an exception of OSM to OGC (which is used by PostgreSQL), see quote below.
I'm surprised it's not flagged as invalid through (in internal table multipolygons I mean, not in OSM). Makes me wonder how it's build internally, e.g. what the MP looks like according to Osmoses internal table.

[wiki](https://wiki.openstreetmap.org/wiki/Relation:multipolygon#Valid_multipolygon_conditions):
> Generally, the multipolygon relation can be used to build multipolygons in compliance with the OGC Simple Feature standard. Anything that is not a valid multipolygon according to this standard (e.g., polygons with intersecting rings) should also be considered an invalid multipolygon relation, **with the notable exception of touching inner rings**

**Comment by sekerob** [comment]
Notably, it was commented 'at me' on the community forum that it's no longer needed *for OSM* to encapsulate touching sides inner areas in a single ring, even got a PM from an 8x blocked mapper who's active in France to stop this method (is textbook) of drawing these *tagless* inner rings around the side sharing inner areas. In the particular case Inspector went all wild as there were about 30 'touching' to include an area inside which then gets rendered same as what's the main MP.... so convenient it was too an excessive large farmland MP relation. I replaced it with a single 'inner' ring and removed the 30 individual from the relation and all quieted down. Told the forum commenter I'd stick to me guns and continue textbook. The PM commenter I ignored. It's only these cases mapped by others which flare up using this 'every single item made inner, regardless'. Even areas that share a side with the MP outline are roled inner, and that Inspector does not like either.

v.v. your quote and the wonder how it looks like according to Osmoses internal table.... I'll leave this one for your case study. I've reported and passed the beer to you ;o).

NB Inspector flags these 'poorly' constructed MPs and observed the few who'd fix these as their seemingly sole mapping activity within 24 hours of showing up have started to ignore (not fix) these when there's lots of buildings inside the farming lands areas... See now several nearing 2 weeks and not being actioned up north.

**Comment by Famlam** [comment]
This is how the relation looks to Osmose. (Obtained via a select on `poly` for `multipolygons` with this id)
[r3211922](https://github.com/osm-fr/osmose-backend/files/14030769/r3211922.zip) (Unzipped it can be loaded in JOSM)

Both inners (way 61773481 and 826931440) are missing from the multipolygon.

**Comment by sekerob** [comment]
Fixed the one in France, hopefully it's now properly complete which we'll know tomorrow. Got you a substitute instead in Italy to test drive to analyzers on. Much more seriously borged because of the failure to understand side sharing/touching inner areas, multiple here, many buildings. Not going to touch, nor have others in the last 2 weeks, Been sitting in the OSM Inspector cross-hairs as the 'known' problem is not the only with that relation to include areas given the inner role when sitting against the MP outline, duplicate segments, wrong role(s), there may be more lurking as some problems wont show until fixing top top set.

https://www.openstreetmap.org/relation/10452441

**Comment by sekerob** [comment]
A new variant popped up this morning. A single inner area to farmland outlining a group of orchards, meadow and residential i.e. no touching sides inner relations, yet a building on the residential area (Semilly) is flagged as being on farmland:

source code 1310/level 3: https://osmose.openstreetmap.fr/en/byuser/SekeRob?username=SekeRob&source=436362
relation of farmland: https://www.openstreetmap.org/relation/14498191#map=14/48.2570/5.4852
building: https://www.openstreetmap.org/way/652699497#map=16/48.2597/5.4606

![image](https://github.com/osm-fr/osmose-backend/assets/23308504/4a6121b4-1dc7-4aa5-a508-41114917a62d)
![image](https://github.com/osm-fr/osmose-backend/assets/23308504/95e64629-5b74-4305-b359-318021bad525)
![image](https://github.com/osm-fr/osmose-backend/assets/23308504/18e58128-d8cc-482d-9297-085694ebda7f)

I've added tome inner farmyards to the relation today and tweaked the outline, but left the subject object untouched.

**Comment by Famlam** [comment]
Regarding the latter: apparently it was related to what you changed, because now that error doesn't exist anymore.
The first error (two touching inner polygons) is a difficult one to solve...

**Comment by sekerob** [comment]
Wellll, another building popped up in the that same Semilly residential area this morning with a pin telling it's on farmland when it is 100% excluded area. Little doubt tomorrow another random building will be selected, There's a 100+ estim to pick from meaning one could hit the false positive X 100+ times, 1 a day, and a new one will show up the next day until stock has been exhausted.

The inner ring containing the residential area according the history has not been touched for over a year https://www.openstreetmap.org/way/63865581 just added bits on the farmland and made those additional inner members so it should not in any way have affected the underlying problem within that larger Semilly inner ring. I have a suspicion since a farmyard with buildings shares a side within the inner ring with the residential area it ticks off the analyzer.  Not a bug but a BUG, maybe in the plural. 

 I'll post the link tomorrow when it shows and hope not to forget to not mouse it to heck.

**Comment by sekerob** [comment]
As expected, same farmland relation, different building in same Semilly residential area shttps://osmose.openstreetmap.fr/en/map/#item=1310&zoom=17&lat=48.258577&lon=5.462297&level=3&tags=&fixable=&issue_uuid=4cc04885-f9dd-5b8e-fe2d-14c246192bd7
https://osmose.openstreetmap.fr/en/byuser/SekeRob?username=SekeRob&source=436362

**Comment by Famlam** [comment]
The majority of these issues will likely be gone when the landuses are becoming more detailed in that region (e.g., split at highways, because there's no crops growing on the highways - I hope). However I am puzzled why the latter ("Semilly-case") is broken: the multipolygon is parsed correctly ([attached the poly as parsed by Osmose](https://github.com/osm-fr/osmose-backend/files/14151365/2138_secondissue.zip)).

So to summarize this issue so far, as it's now about two separate things in one discussion:
- One issue is that two connecting inner polygons are dropped by the current multipolygon parsing, discussed up to https://github.com/osm-fr/osmose-backend/issues/2138#issuecomment-1907086367 . This issue is clear, just difficult to solve.
- The other issue (discussions after that comment) is a separate bug where for some reason some ways in inners are still parsed as if they're not in the inner. I have no clue at the moment why this is the case, and a local simplified test case didn't allow me to reproduce the issue.

**Comment by sekerob** [comment]
For the bad news, Bug 2 is maybe BUG 2. This morning, totally hands off, the farmland around Semilly was flagged again, but now with a building on a farmyard named Caurois which is totally stand alone, no side sharing anything, a simple inner polygon

https://osmose.openstreetmap.fr/en/byuser/SekeRob?username=SekeRob&source=436362
https://www.openstreetmap.org/relation/14498191

![image](https://github.com/osm-fr/osmose-backend/assets/23308504/deb00fb1-913b-43da-9498-13d62ca55dad)
![image](https://github.com/osm-fr/osmose-backend/assets/23308504/1a35fb30-6ea8-47f6-b04e-8860200e24ed)

I'll declare this zona non grata, something more is wrong here, maybe along the line of dropping the ball on the first encounter of a building deemed on farmland and stopping any further rule checking, then just pick any one building within the MP area.

**Comment by sekerob** [comment]
FWIW, case 1 is (still) on occasion encountered in other farming lands places. Case 2, unless a code change could explain this, the "Semilly" problem and randomly flagging of a building within that inner polygon area and e.g. farmyards with mapped building which were correctly given an inner role to the farmland MP went away the moment I split the farmland along a traversing primary road. By splitting I mean a road corridor thru the farmland relation  That I did about 10-12 days ago. From then on only buildings not yet having an 'inner' area mapped around with inner role, both sides of the road, were being flagged, 1 a day of course until yesterday when I did the last one, intentionally doing only one/few at the time to see if this was a chance or persistent state. Well as noted, the MP was not flagged this morning.

--//--
